src/app/server.rs

```diff
@@ -521,106 +507,101 @@ fn get_env(
-        if let Some(environment) = db::read_sql_env(&state.db_path, env, sha).unwrap_or(None) {
-            debug!("Found environment in the db {} {}", env, sha);
-            Ok(insert_into_env_cache(state, &key, environment))
-        } else {
-            let _write_lock = state.write_lock.lock();
-
-            //Double check if the cache now contains the env we are looking for
-            if let Some(environment) = db::read_sql_env(&state.db_path, env, sha).unwrap_or(None) {
-                register_cache_hit(state, &key);
-                debug!("Avoided git lock for config lookup: {}", key);
-                return Ok(Arc::new(environment));
+        // This locks mutating the git repo
+        let _write_lock = match state.write_lock.try_lock_for(Duration::from_secs(20)) {
+            Some(lock) => lock,
+            None => {
+                warn!(
+                    "Unable to acquire write lock in a timely manner: {} {}",
+                    sha, env
+                );
+                return Err(HoganError::InternalTimeout.into());
             }
+        };
 
-            register_cache_miss(state, &key);
+        //Since waiting for the lock may take unknown time double check if the cache now contains the env we are looking for
+        if let Some(environment) = db::read_sql_env(&state.db_path, env, sha).unwrap_or(None) {
+            register_cache_hit(state, &key);
+            debug!("Avoided git lock for config lookup: {}", key);
+            return Ok(Arc::new(environment));
+        }
 
-            let sha = state
-                .config_dir
-                .refresh(remote, Some(sha), state.allow_fetch)?;
+        register_cache_miss(state, &key);
 
-            let filter = match hogan::config::build_env_regex(env, Some(&state.environment_pattern))
-            {
-                Ok(filter) => filter,
-                Err(e) => {
-                    warn!("Incompatible env name: {} {:?}", env, e);
-                    //In an error scenario we'll still try and match against all configs
-                    state.environments_regex.clone()
-                }
+        let sha = state
+            .config_dir
+            .refresh(remote, Some(sha), state.allow_fetch)?;
+
+        let filter = match hogan::config::build_env_regex(env, Some(&state.environment_pattern)) {
+            Ok(filter) => filter,
+            Err(e) => {
+                warn!("Incompatible env name: {} {:?}", env, e);
+                //In an error scenario we'll still try and match against all configs
+                state.environments_regex.clone()
+            }
+        };
+        if let Some(environment) = state
+            .config_dir
+            .find(filter)
+            .iter()
+            .find(|e| e.environment == env)
+        {
+            if let Err(e) = db::write_sql_env(&state.db_path, env, &sha, environment) {
+                warn!("Unable to write env {} {}::{} to db {:?}", key, sha, env, e);
```

Comment:
shouldnt this now be `key, env, sha, e` (or use the format_key function to be consistent